src/ringct/rctSigs.cpp

```diff
@@ -1600,17 +1602,21 @@ namespace rct {
 
       ::do_serialize(ar, const_cast<rctSig&>(rv));
 
-      crypto::hash h;
-      cryptonote::get_blob_hash(ss.str(), h);
+      std::string ser_sig_and_txhash = ss.str();
+      const crypto::hash tx_hash = cryptonote::get_transaction_hash(tx);
+      ser_sig_and_txhash.append(tx_hash.data, sizeof(crypto::hash));
```

Comment:
Wow ok so this does require two allocations without some fancy stringstream ops to copy the underlying stringstream buffer, that's too bad.

You need to include a domain separator:

```
cryptonote_config.h

const constexpr char HASH_KEY_CONTEXTUAL_TX_VALIDATION_ID[] = "contextual_tx_validation_id";
```

```
in this file:

std::string ser_sig_and_txhash;
ser_sig_and_txhash.reserve(sizeof(config::HASH_KEY_CONTEXTUAL_TX_VALIDATION_ID) + ss.tellp() + sizeof(crypto::hash));
ser_sig_and_txhash.append(config::HASH_KEY_CONTEXTUAL_TX_VALIDATION_ID);
ser_sig_and_txhash.append(ss.str());
ser_sig_and_txhash.append(tx_hash.data, sizeof(crypto::hash));
```

Reply:
What's the point of a domain separator in a data structure that's not mixed with any other hashes anywhere? At this point it's "do everything by the book" just for the sake of doing by the book. Domain separator is unnecessary here.

Reply:
Doesn't matter if it's unnecessary, do it anyway.

Reply:
Explain why? The code gets fatter and more complicated. I'm against this change.

Reply:
It's a policy for higher-quality cryptographic code.

Reply:
I know what it is and why. Anyway, after the latest commit it's much cleaner and just 1 line of code, so good enough.

Reply:
The transcript utility I wrote for seraphis requires a domain separator in the constructor, to make it extra clear when **you're doing it wrong**.

Reply:
Yes, check the latest commit. Should be a good compromise. All allocations happen in `std::stringstream`, and we use best practices of domain separation.